crates/circuit/src/dag_circuit.rs

```diff
@@ -2813,25 +2813,76 @@ def _format(operand):
+        let node2 = node2.node.unwrap();
+
+        // Check that both nodes correspond to operations
+        if !matches!(self.dag.node_weight(node1).unwrap(), NodeType::Operation(_))
+            || !matches!(self.dag.node_weight(node2).unwrap(), NodeType::Operation(_))
+        {
+            return Err(DAGCircuitError::new_err(
+                "Nodes to swap are not both DAGOpNodes",
+            ));
+        }
+
+        // Gather all wires connecting node1 and node2.
+        // This functionality was extracted from rustworkx's 'get_edge_data'
+        let wires: Vec<Wire> = self
+            .dag
+            .edges(node1)
+            .filter(|edge| edge.target() == node2)
+            .map(|edge| edge.weight().clone())
+            .collect();
+
+        if wires.is_empty() {
+            return Err(DAGCircuitError::new_err(
+                "Attempt to swap unconnected nodes",
+            ));
+        };
+
+        // Closure that finds the first parent/child node connected to a reference node by given wire
+        // and returns relevant edge information depending on the specified direction:
+        //  - Incoming -> parent -> outputs (parent_edge_id, parent_source_node_id)
+        //  - Outgoing -> child -> outputs (child_edge_id, child_target_node_id)
+        // This functionality was inspired in rustworkx's 'find_predecessors_by_edge' and 'find_successors_by_edge'.
+        let directed_edge_for_wire = |node: NodeIndex, direction: Direction, wire: &Wire| {
+            for edge in self.dag.edges_directed(node, direction) {
+                if wire == edge.weight() {
+                    match direction {
+                        Incoming => return Some((edge.id(), edge.source())),
+                        Outgoing => return Some((edge.id(), edge.target())),
+                    }
+                }
+            }
+            None
+        };
+
+        // Vector that contains a tuple of (wire, edge_info, parent_info, child_info) per wire in wires
+        let relevant_edges = wires
+            .iter()
+            .rev()
+            .map(|wire| {
+                (
+                    wire,
+                    directed_edge_for_wire(node1, Outgoing, wire).unwrap(),
+                    directed_edge_for_wire(node1, Incoming, wire).unwrap(),
+                    directed_edge_for_wire(node2, Outgoing, wire).unwrap(),
+                )
+            })
+            .collect::<Vec<_>>();
+
+        // Iterate over relevant edges and modify self.dag
+        for (wire, (node1_to_node2, _), (parent_to_node1, parent), (node2_to_child, child)) in
+            relevant_edges
+        {
+            self.dag.remove_edge(parent_to_node1);
+            self.dag.add_edge(parent, node2, wire.clone());
+            self.dag.remove_edge(node1_to_node2);
+            self.dag.add_edge(node2, node1, wire.clone());
+            self.dag.remove_edge(node2_to_child);
+            self.dag.add_edge(node1, child, wire.clone());
```

Comment:
After updating the branch and fixing the merge conflicts, the compiler started complaining about `wire` being moved and having to clone it. I am not sure what changed to motivate that.

Reply:
This is because `Wire` no longer implements `Copy` as of the latest `oxidize-dag` branch since it now needs to store a `PyObject` (which itself doesn't implement `Copy`) to represent classical variable wires.

Once classical expressions / variables are ported to Rust (which will happen sometime after the `DAGCircuit` oxidization is complete), we can likely make `Wire` implement `Copy` again, and then you'd be able to do `*` to copy these again.